fortanix-vme/fortanix-vme-runner/src/lib.rs

```diff
@@ -186,79 +218,87 @@ impl ConnectionKey {
     }
 }
 
-pub struct Server {
-    command_listener: Mutex<VsockListener>,
-    /// Tracks information about TCP sockets that are currently listening for new connections. For
-    /// every TCP listener socket in the runner, there is a vsock listener socket in the enclave.
-    /// When the enclave instructs to accept a new connection, the runner accepts a new TCP
-    /// connection. It then locates the ListenerInfo and finds the information it needs to set up a
-    /// new vsock connection to the enclave
-    listeners: RwLock<FnvHashMap<VsockAddr, Arc<Mutex<Listener>>>>,
-    connections: RwLock<FnvHashMap<ConnectionKey, ConnectionInfo>>,
+pub struct ClientConnection<'de> {
+    sender: VsockStream,
+    reader: StreamDeserializer<'de, IoRead<VsockStream>, Request>,
 }
 
-impl Server {
-    fn log_communication(src: &str, src_port: u32, dst: &str, dst_port: u32, msg: &str, arrow: Direction, prot: &str) {
+impl<'de> ClientConnection<'de> {
+    pub fn new(stream: VsockStream) -> Self {
+        // Mimic splitting the stream in a read and send part
+        let sender = unsafe { VsockStream::from_raw_fd(stream.as_raw_fd()) };
```

Comment:
`tokio::net::TcpStream::split` does exactly this, in an elegant way. When we move to Async such support may be added to `vsock` as well. It used to work without this my enabling read/write on top of a `&VsockStream`.